src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs

```diff
@@ -70,22 +72,31 @@ public JsonClassInfo ElementClassInfo
 
         public void UpdateSortedPropertyCache(ref ReadStackFrame frame)
         {
-            // Check if we are trying to build the sorted cache.
-            if (frame.PropertyRefCache == null)
-            {
-                return;
-            }
+            Debug.Assert(frame.PropertyRefCache != null);
 
-            List<PropertyRef> newList;
-            if (_propertyRefsSorted != null)
-            {
-                newList = new List<PropertyRef>(_propertyRefsSorted);
-                newList.AddRange(frame.PropertyRefCache);
-                _propertyRefsSorted = newList.ToArray();
-            }
-            else
+            lock (_cacheLock)
             {
-                _propertyRefsSorted = frame.PropertyRefCache.ToArray();
+                if (_propertyRefsSorted != null)
+                {
+                    List<PropertyRef> replacementList = new List<PropertyRef>(_propertyRefsSorted);
+                    Debug.Assert(replacementList.Count <= PropertyNameCountCacheThreshold);
+
+                    // Verify replacementList will not become too large. This could occur in race conditions.
```

Comment:
What race conditions (given that we're now under a lock)?

Reply:
Is it the term "race condition"? Yes the race condition is now protected against malformed writes.

Multiple threads each build up their local temporary list which is protected to not surpass 64 including this shared cache; however that check for 64 is not protected by a lock against the shared cache. Then when this method is later called to append a given thread's temporary list to the shared cache, the >64 must be done again within a lock.